File: leetcode/03-Longest-Substring-Without-Repeating-Characters/python/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        result = 0
        start = 0
        map = {}
        for end in range(len(s)):
            currentChar = s[end]
            if currentChar in map and start <= map[currentChar]:
                # すでに今の文字が出てきていたらstart位置をずらす
                start = map[currentChar] + 1

            # 現在の文字をmapに保存する
            map[currentChar] = end

            currentLength = end - start + 1
            result = currentLength if currentLength > result else result

        return result

File: leetcode/03-Longest-Substring-Without-Repeating-Characters/go/main.go
package main

func lengthOfLongestSubstring(s string) int {
	left := 0
	result := 0
	cache := make(map[string]int)

	for right := 0; right < len(s); right++ {
		currentChar := s[right : right+1]
		if cachedIndex, hasCache := cache[currentChar]; hasCache && cachedIndex >= left {
			left = cachedIndex + 1
		}

		cache[currentChar] = right
		currentLength := right - left + 1
		if currentLength > result {
			result = currentLength
		}
	}

	return result
}
